i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -229,4 +425,573 @@ mod tests {
 
         Ok(())
     }
+
+    #[test]
+    fn test_create_catalog_nested_directories() -> anyhow::Result<()> {
+        let (ctx, _tmp) = create_render_context(&[
+            ("book.toml", "[book]"),
+            (
+                "src/SUMMARY.md",
+                "- [The Foo Chapter](foo.md)\n\
+                \t- [The Bar Section](foo/bar.md)\n\
+                \t\t- [The Baz Subsection](foo/bar/baz.md)",
+            ),
+            (
+                "src/foo.md",
+                "# How to Foo\n\
+                 \n\
+                 The first paragraph about Foo.\n",
+            ),
+            (
+                "src/foo/bar.md",
+                "# How to Bar\n\
+                 \n\
+                 The first paragraph about Bar.\n",
+            ),
+            (
+                "src/foo/bar/baz.md",
+                "# How to Baz\n\
+                 \n\
+                 The first paragraph about Baz.\n",
+            ),
+        ])?;
+
+        // Using a depth of 0 to include all messages in a single template file.
+        let catalogs = create_catalogs(&ctx, 0)?;
+        let catalog = &catalogs[&default_template_file()];
+
+        for msg in catalog.messages() {
+            assert!(!msg.is_translated());
+        }
+
+        let expected_message_tuples = vec![
+            ("src/SUMMARY.md:1", "The Foo Chapter"),
+            ("src/SUMMARY.md:2", "The Bar Section"),
+            ("src/SUMMARY.md:3", "The Baz Subsection"),
+            ("src/foo.md:1", "How to Foo"),
+            ("src/foo.md:3", "The first paragraph about Foo."),
+            ("src/foo/bar.md:1", "How to Bar"),
+            ("src/foo/bar.md:3", "The first paragraph about Bar."),
+            ("src/foo/bar/baz.md:1", "How to Baz"),
+            ("src/foo/bar/baz.md:3", "The first paragraph about Baz."),
+        ];
+
+        let message_tuples = catalog
+            .messages()
+            .map(|msg| (msg.source(), msg.msgid()))
+            .collect::<Vec<(&str, &str)>>();
+
+        for message_tuple in expected_message_tuples {
+            assert!(message_tuples.contains(&message_tuple));
+        }
```

Comment:
I think you can assert something stronger: you know the order in which you walk the tree of chapters, so you know the order of the messages in `catalog`.

The order is important since it directly affects the lines written into the POT files. So it is important that it is not accidentally randomized via iteration over the hashmap.

Basically, I suggest

```rust
assert_eq!(
    catalog.messages().map(|msg| msg.source()).collect::<Vec<_>>(),
    vec![
        "src/SUMMARY.md:1",
        ...
    ]
)
```

While writing this, I figured that we might not need to look at the `msgid` fields at all — there is probably another test which checks them?

Reply:
Good point about enforcing that messages are in order. And yes there is another test that deals with the `msgid` fields, so we don't need that here

Done!